### PR TITLE
Fix unordered is equal

### PR DESCRIPTION
**Pull request by ngeiswei**
Fix for issue #920 

I'm rather confident so I'll merge it tomorrow morning European time, but @linas you might want to have a look at it.

### CODE REVIEW

**Comment by ngeiswei** [comment]
Don't merge that just yet there is a subtle bug that would occur if an unordered link contains duplicated atoms.

File: opencog/atoms/base/Handle.h [hunk 1]
```diff
@@ -263,6 +268,17 @@ struct content_based_atom_ptr_less
     }
 };
 
+struct content_based_handle_less
+{
+    bool operator()(const Handle& hl, const Handle& hr) const
+    {
+        return Handle::content_based_atoms_less(hl.const_atom_ptr(),
+                                                hr.const_atom_ptr());
+    }
```
Comment (linas) [inline opencog/atoms/base/Handle.h]: Why are you passing Atom* here? why not just pass in the handles themselves?  Its kind-of dangerous to pass around raw, naked Atom* -- which is why the API mostly doesn't have Atom*'s in it anywhere -- except for this new usage.

File: opencog/atoms/base/Link.cc [hunk 2]
```diff
@@ -137,21 +137,69 @@ bool Link::operator==(const Atom& other) const
...
+        it = lhs.begin(),
+        other_it = rhs.begin();
+
+    for (; it != lhs.end(); it++, other_it++)
+    {
+        if ((*it)->getType() != (*other_it)->getType())
+            return false;
+
+        if ((*it)->isNode())
+        {
```
Comment (linas) [inline opencog/atoms/base/Link.cc]: This is not required -- PrototAtom has a virtual `operator==()` in it already. So it should be enough to just say `*lhs == *rhs`  and be done with it. You can skip the type comparison, above, also.

The core issue turns out to be that NodeCast and LinkCast are insanely expensive.

File: opencog/atoms/base/Link.cc [hunk 3]
```diff
@@ -137,21 +137,69 @@ bool Link::operator==(const Atom& other) const
...
+            LinkPtr ol(LinkCast(other_outgoing[i]));
+            if (*tl != *ol) return false;
+        }
+    }
+    return true;
+}
+
+bool Link::content_based_equal(const ContentBasedOrderedHandleSet& lhs,
+                               const ContentBasedOrderedHandleSet& rhs) const
+{
```
Comment (linas) [inline opencog/atoms/base/Link.cc]: I mean, maybe if I studied it hard, I could figure it out, but then I will forget, and the next person won't understand this, either...

File: opencog/atoms/base/Link.cc (hunk 2 above)
Comment (ngeiswei) [inline opencog/atoms/base/Link.cc]: I've merely copied/pasted this code from the previous `Link::operator==`. But before I did that I tried to return `lhs[i] == rhs[i]` and for some reason it broke the unit tests, so I concluded that there's must be a reason for that and let it there. I'll investigate though.

Comment (ngeiswei) [inline opencog/atoms/base/Link.cc]: I found the problem, Handle::operator== directly compare the pointers, see https://github.com/opencog/atomspace/blob/master/opencog/atoms/base/Handle.h#L153 this is wrong too. Maybe I should fix it as well with a similar fix, that is if one of them have undefined UUID they might have different pointers but same content.

**Comment by ngeiswei** [comment]
I'm merging. I have let `Handle::operator!=`, I fear it is buggy but I prefer to address when a problems will appear due to it.

File: opencog/atoms/base/Link.cc (hunk 2 above)
Comment (linas) [inline opencog/atoms/base/Link.cc]: Ugh.  I made some progress yesterday in removing UUID's. (They will remain in SQL, and in anything else that needs them -- e.g. the rest server, the zmq server)

I retract the comment that started this conversation: it *should* be enough to just call the overloaded operator== but it turns out that this is difficult to implement correctly, for C++-ish reasons, so changing this is probably a low priority. The code works, for now.

**Pull request merged**